Optimize weight redistributor by using simplified "single-box" packing algorithm

colinmollenhour opened the pull request:
This should reduce the amount of work performed to rebalance weight when there are a very large number of boxes in particular.

colinmollenhour commented:
@dvdoug Can you do a quick review? I was expecting this to have the same output, just faster but somehow quite a few of the efficiency tests have regressed. The logic is just that when doing the weight redistribution since the return value of `doVolumeRepack` is only useful if it contains a single packed box, then the packer doesn't need to do as much work to determine which box best fits the full set of items, it can just iterate over boxes smallest to largest and stop when the first one fits all items. Is this flawed somehow?

dvdoug commented on src/ItemList.php:
```diff
@@ -175,6 +175,15 @@ public function hasConstrainedItems(): bool
+    {
+        $volume = 0;
+        foreach ($this->list as $item) {
+            $volume += $item->getDepth() * $item->getWeight() * $item->getLength();
```
I think you meant `* $item->getWidth()` 😉 . Don't forget to add a test.

(width, length, depth) is the ordering I try and stick to

colinmollenhour replied:
Ahh, good catch! I removed the method since I see you've done something similar in the new `getBoxList` method to stay consistent. Let me know if you'd prefer it to be added back to reduce code duplication.

colinmollenhour commented:
The build is passing now. Let me know what you think. Thanks!

dvdoug commented on src/BoxList.php:
```diff
@@ -59,11 +59,11 @@ public static function compare($boxA, $boxB): int
             return $volumeDecider;
         }
+
+        $emptyWeightDecider = $boxA->getEmptyWeight() <=> $boxB->getEmptyWeight(); // with smallest empty weight
```
Good spot, there's definitely an optimisation to be had here. But also because of this, it's not possible to tell from looking at the timings on the Travis run what the performance impact of the other change is.

So what I'm going to do is steal this bit out of the PR and commit it to master. Then the comparison for the rest should be more straightforward.

colinmollenhour replied:
Sounds good. Just FYI I'm testing with ~170 boxes so while the unit tests might show a small improvement I think the improvement for cases like mine will be much more significant. I don't have solid numbers yet, was just looking for some low hanging fruit.

dvdoug replied:
170! 😱 

I love a good torture test, it would be great to have a sample of that in the test suite if that would be OK? (`tests/PackerTest.php` would be the right placed to put it). Just a single example of the typical items in a packing and the 170 boxes.

Apart from obviously helping evaluate this PR in particular, it would also ensure that performance on that kind of scenario stays good going forward and doesn't regress.

colinmollenhour replied:
I'll work on adding such a test. Any other thoughts in this PR as it is written currently?

colinmollenhour commented:
Rebased and now uses the new `singlePassMode` flag.

colinmollenhour commented:
I rebased on the latest master and added a test case with 135 boxes and this optimization only improves performance by about 1% so not as much as I expected. Let me know if you want me to submit separate PRs for each commit and thoughts on the optimization.

dvdoug commented:
> 
> 
> I rebased on the latest master and added a test case with 135 boxes and this optimization only improves performance by about 1% so not as much as I expected. Let me know if you want me to submit separate PRs for each commit and thoughts on the optimization.

Awesome, thanks! I'll take a proper look (probably tomorrow) and run both the testcase and the change through the profiler. I suspect the 1% is probably about right, as this is effectively a slightly more optimised of https://github.com/dvdoug/BoxPacker/commit/f0ef0d98da6c8dd2f26baae1c9e21f17e890d781 which I added at the same time you had the idea for this - without it, I think this would have been a big win.

dvdoug commented:
If the data shows what I suspect, then maybe an extra param on `getBoxList()` to exclude the "non-preferreds" (will need a different name) might be better to avoid 2 methods that are very, very similar.

dvdoug commented:
Hi @colinmollenhour 

I spent a bit of time profiling this, and the impact came out to ~0.4% on that torture test, which matches up with the timing shown of the Travis builds. The callgraph showed that this PR didn't reduce the number of packing iterations on that testcase (i.e. the number of calls from `Packer` to `VolumePacker` was identical), the difference in runtime was purely inside `Packer`, where the existing code is doing slightly more work the the single box optimised version you added.

However, adding that near-duplicate method does increase complexity by quite a bit (see https://scrutinizer-ci.com/g/dvdoug/BoxPacker/inspections/1c511f55-1b81-4de7-9231-af83c8c47d67) and therefore I don't think the tradeoff is worth it.

I can see where excluding box(es) from the weight redistribution repack might help on other cases though, so I've added a variation on your idea.

So sorry I can't merge the code on this one, but thank you for the idea and the torture test 😃

colinmollenhour commented:
Thanks @dvdoug for taking the time to incorporate the ideas for the small improvement. I appreciate the thorough code review process and standards so no need for apologies!

Any chance you will tag a release soon?

dvdoug commented:
Yes, should be this week
